Added modal reactors documentation (#62)
lf-lang/website-lingua-franca: packages/documentation/copy/en/topics/Modal Models.md

State variables are not reset automatically to their initial conditions because it is idiomatic for reactors to allocate resources or initialize subsystems (e.g., allocate memory or sockets, register an interrupt, or start a server) in reactions triggered by the `startup`, and to store references to these resources in state variables.
If these were to be automatically reset, those references would be lost.

On the other hand, if a mode has been active prior and is then re-entered via a `continue` transition, no reset is performed.
Events originating from timers, scheduled actions, and delayed connections are adjusted to reflect a remaining delay equal to the remaining delay recorded at the instant the mode was previously deactivated.
As a consequence, a mode has a notion of local time that elapses only when the mode is active.

### Local Time

From the perspective of timers and actions, time is suspended when a mode is inactive.
This also applies to indirectly nested reactors within modes and connections with logical delays, if their source lies within a mode.

<img alt="Illustration of local time (model)" src="../../../../../img/modal_models/local_time.svg" width="400"/>

> The given LF model illustrates the different characteristics of local time affecting timers and actions in the presence of the two transition types.

Reviewer: ```suggestion
The above LF model illustrates the different characteristics of local time affecting timers and actions in the presence of the two transition types.
```

It consists of two modes `One` (the initial mode) and `Two`, both in the `Modal` reactor.
The `next` input toggles between these modes and is controlled by a reaction at the top level that is triggered by the timer `T`.
After one second, a mode switch is triggered periodically with a period one second.
Each mode has a timer `T1`/`T2` that triggers a reaction after an initial offset of 100 msec and then periodically after 750 msec.
This reaction then schedules a logical action with a delay of 500 msec (the actual target code does not add an additional delay over the minimum specified).
This action triggers the second reaction, which writes to the output `out`.
The main difference between the modes is that `One` is entered via a history transition, continuing its behavior, while `Two` is reset.
(Continue behavior is indicated by an "H" on the transition edge because it enters into the entire history of the mode.)

<img alt="Illustration of local time (trace)" src="../../../../../img/modal_models/local_time_trace.svg" width="600"/>

> This is the execution trace of the first 4 seconds of this program.

Reviewer: ```suggestion
Above is the execution trace of the first 4 seconds of this program.
```

Below the timeline is the currently active mode and above the timeline are the model elements that are executed at certain points in time, together with indicating triggering and their relation through time.
For example, at 100 msec, the initial offset of timer `T1` elapses, which leads to the scheduling of the logical action in this mode.
The action triggers the reaction 500 msec later, at 600 msec, and thus causes an output.
The timing diagram illustrates the different handling of time between history transitions and reset transitions.
Specifically, when mode `One` is re-entered via a history transition, at time 2000 msec, the action triggered by `T1` before, at time 850 msec, resumes.
In contrast, when mode `Two` is re-entered via a reset transition, at time 3000 msec, the action triggered by `T2` before, at time 1850 msec, gets discarded.

<img alt="Illustration of local time (plot)" src="../../../../../img/modal_models/local_time_plot.svg" width="300"/>

> This plot illustrates the relation between global time in the environment and the localized time for each timer in the model.

Reviewer: ```suggestion
The above plot illustrates the relation between global time in the environment and the localized time for each timer in the model.
```

Since the top-level reactor `TimingExample` is not enclosed by any mode, its time always corresponds to the global time.
Mode `One` is the initial mode and hence progresses in sync with `TimingExample` for the first second.
During inactivity of mode One the timer is suspended and does not advance in time.
At 2000 msec it continues relative to this time.
`T2` only starts advancing when the mode becomes active at 1000 msec.
The reentry via reset at 3000 msec causes the local time to be reset to zero.

### Startup and Shutdown

**This behavior may be subject to change in the future**

A challenge for modal execution is the handling `startup` and `shutdown` behavior.
These are commonly used for managing memory for state variables, handling connections to sensors or actuators, or starting/joining external threads.
If reactions to these triggers are located inside modes they are subject to a special execution regime.

First, `startup` reactions are invoked at most once at the first activation of a mode.

> Second, `shutdown` reactions are executed when the reactor shuts down, ***irrespective*** of mode activity, but we only if the enclosing modes have been activate at least once.

Reviewer: ```suggestion
Second, `shutdown` reactions are executed when the reactor shuts down, ***irrespective*** of mode activity, but only if the enclosing modes have been activated at least once.
```

Hence, every startup has a corresponding shutdown.
Third, as mentioned before, the new `reset` trigger for reactions can be used, if a startup behavior should be re-executed if a mode is entered with a reset transition.

Note that this may have unexpected consequences:

* Startup behavior inside modes may occur during execution and not only at program start.
* Multiple shutdown reactions may be executed, bypassing mutual exclusion of modes.

> * Reactors that are design without consideration of modes and use only `startup` (not `reset`) to trigger an execution chain, may not work in modes and cease to function if re-entered with a reset.

Reviewer: ```suggestion
* Reactors that are designed without consideration of modes and use only `startup` (not `reset`) to trigger an execution chain, may not work in modes and cease to function if re-entered with a reset.
```
